### ISSUE TITLE
There should be a way to set a custom deserializer.

### ISSUE THREAD

**Issue by louy** [lytics/ember-data.model-fragments#95]
`FragmentTransform.deserialize()` is not implemented yet.

https://github.com/lytics/ember-data.model-fragments/blob/77a9a4eefde1bbaf20cb86df4fe4550fc9f77deb/packages/model-fragments/lib/fragments/transform.js#L21

Currently, the only way to do a custom deserialize is to do it in the model Serializer. This is due to the limitations of `DS.Transform`.

**Comment by slindberg** [comment]
I think you may be misunderstanding `DS.FragmentTransform`'s job. Fragment properties behave like attributes (unlike relationships, which are handled specially), so when deserializing the fragment attribute, the transform's job is simply to return an _already deserialized_ fragment. If you wish to provide custom deserialization logic for the fragment, you can create a [specific serializer](https://github.com/lytics/ember-data.model-fragments#serializing) for your fragment class. They behave exactly as model serializers do.

I hope this helps, if I've misunderstood, feel free to re-open.

**Comment by louy** [comment]
You're right, but that's the exact issue.
Take a look at the link I provided. The transformer class is the one that calls the Serializer, and it only does that on `serialize`. So `extract` is never called. There's also a comment that says `@TODO` in the same file mentioning this :)

**Comment by louy** [comment]
By the way it seems I can't reopen the issue.
I'm not sure if I'm missing something here but I've tried `extract`, `extractArray` and `extractSingle` and none of them were ever called. If I'm not mistaken then please reopen this issue.

**Comment by slindberg** [comment]
@louy Ah yes, you are correct that `extract` serializer methods are not used, I need to make this clear in the readme. I understand now what you meant by

> FragmentTransform.deserialize() is not implemented yet.

I fully agree that the ideal place for payload extraction is in the fragment transform, which would simply be a matter of calling `serializer.extract`. However, [all type information is lost during the call to `deserialize`](https://github.com/emberjs/data/blob/e97d2dba8cc099cacbe09d5e01454453bf2797ef/packages/ember-data/lib/serializers/json-serializer.js#L121). That's what that `TODO` references, and why it's still outstanding. It's only happenstance that the fragment can get serialized in the transform, due to to the object being passed in being a `Snapshot` instance with related type data, unlike the POJO passed to `deserialized`.

The two options for getting the model type/serializer in the transform are overriding a not-so-private serializer method completely, or registering a new transform class for every model fragment, neither of which I felt were safe/stable enough to attempt. For now, you're right that you'll have to do any `extract` logic in parent model's `normalize` method. I'll re-open the issue to track it (weird that you couldn't, as the creator?).

**Comment by louy** [comment]
Exactly. I have opened [an issue on Ember-data](https://github.com/emberjs/data/issues/3111). Let's see if they will do something about it.
By the way, I can [only](http://stackoverflow.com/a/21333938/1644422) reopen issues if I closed them *myself*.

**Comment by slindberg** [comment]
This should be resolved when #109 lands. You'll be able to specify your own transform:

```javascript
// app/transforms/fragment.js
export default DS.FragmentTransform.extend({
  deserialize: function(data) {
    // Do some special handling of the raw JSON attribute

    return this._super(data);
  }
});
```

Unfortunately you'll also have to go through this special incantation:

```javascript
import FragmentTransform from `../transforms/fragment`;

// app/initializers/register-transforms.js
export function initialize(/* container, application */) {
  application.register('transform:fragment', FragmentTransform);
}

export default {
  name: 'register-transforms',
  initialize: initialize
};
```
